### ISSUE TITLE
JDK 21 Support

### ISSUE THREAD

**Issue by overheadhunter** [purejava/appindicator-gtk3-java#14]
### Please agree to the following

- [X] I have searched [existing issues](https://github.com/purejava/appindicator-gtk3-java/issues?q=) for duplicates

### Summary

Build the project with/for JDK 21

### Motivation

Due to Panama being a preview feature, classes need to be compiled with `--enable-preview`:

https://github.com/purejava/appindicator-gtk3-java/blob/26f00bf50e33237b53a288f08909e86a2ee80c12/pom.xml#L130-L141

This results in class files not being version `64` (Java 20) but rather `64.65535`. Attempting to load those classes with newer Java versions will fail.

In other words: Classes compiled with `--enable-preview` are bound to the one exact Java version they are compiled with, as there are no compatibility guarantees for preview features.

Downstream projects can not move to JDK 21 (or would need to temporarily remove the dependency until this is resolved 😉).

### Anything else?

This is clearly a breaking change, so you probably want to release it as a major version

**Comment by purejava** [comment]
I am a little lost @overheadhunter. [Seems](https://github.com/purejava/appindicator-gtk3-java/actions/runs/6112926486), that `SegmentScope` has been removed from Java 21, but I can't find any [documentation](https://download.java.net/java/early_access/jdk21/docs/api/java.base/java/lang/foreign/SegmentScope.html) about that or what class to use instead.

**Comment by overheadhunter** [comment]
What you're looking for is probably `Arena`, which defines both the scope/lifetime as well as acts as an allocator.

[(not really) docs](https://cr.openjdk.org/~mcimadamore/panama/scoped_arenas.html)

**Comment by purejava** [comment]
Thank you very much for your help @overheadhunter. That'll probably it.

I took another turn on the code, and it came to my mind, that the Java bindings need to be regenerated, as the code generated by `jextract` with Java 20 uses `SegmentScope` too:
https://github.com/purejava/appindicator-gtk3-java/blob/13a0758b23caaad3e307c7f2026ba67750d63c58/src/main/java/org/purejava/appindicator/RuntimeHelper.java#L123

I'll have to wait for `jextract` to be based and [released on Java 21](https://jdk.java.net/jextract/). Hopefully, this happens soon after the release of Java 21 this month.

**Comment by overheadhunter** [comment]
I can start a PR and run jextract for you (if you know how, building it from source is easy, but there are some caveats due to bad documentation).

But you have to finish the PR due to my limited understanding of the API.

**Comment by overheadhunter** [comment]
For the sake of completeness, here is how to build jextract:

1. checkout the jdk21 branch
2. install llvm (as linked from jextract's README)

Now here is the important part that the README doesn't tell you:

3. set you JAVA_HOME to an **old** JDK, e.g. `sdk use java 17...`, otherwise gradle won't run on 20/21
4. run the gradle build as shown on the README and pass the path to JDK21 as well as the path to llvm in the corresponsing command line arguments
5. done, the build just takes a few seconds

**Comment by purejava** [comment]
That sounds exactly like something I'd love to give a try. 😃 
With your hints, building `jextract` from source worked very well!
```shell
ralph@fusion jextract % sh ./gradlew -Pjdk21_home=/home/ralph/.sdkman/candidates/java/21.ea.35-open -Pllvm_home=/home/ralph/Downloads/clang+llvm-16.0.0-aarch64-linux-gnu clean verify

> Task :compileJava
Hinweis: Einige Eingabedateien verwenden Vorschaufeatures von Java SE 21.
Hinweis: Wiederholen Sie die Kompilierung mit -Xlint:preview, um Details zu erhalten.

BUILD SUCCESSFUL in 18s
8 actionable tasks: 8 executed
ralph@fusion jextract %
```
```shell
ralph@fusion bin % ./jextract --version
jextract 21
JDK version 21+35-2513
clang version 16.0.0 (http://git.linaro.org/toolchain/jenkins-scripts.git a0f5cfaf8565d1d30567b71dd1f6a1724a6513ac)
ralph@fusion bin %
```
And you were right: `SegmentScope` indeed got replaced by `Arena`. I'll have to create some patches for the different branches of this repo:
```diff
diff src/main/java/org/purejava/appindicator/RuntimeHelper.java org/purejava/appindicator/RuntimeHelper.java
1d0
< // Generated by jextract
2a2
> // Generated by jextract
4,9c4,12
< import org.slf4j.Logger;
< import org.slf4j.LoggerFactory;
< 
< import java.io.File;
< import java.io.IOException;
< import java.lang.foreign.*;
---
> import java.lang.foreign.Linker;
> import java.lang.foreign.FunctionDescriptor;
> import java.lang.foreign.GroupLayout;
> import java.lang.foreign.SymbolLookup;
> import java.lang.foreign.MemoryLayout;
> import java.lang.foreign.MemorySegment;
> import java.lang.foreign.Arena;
> import java.lang.foreign.SegmentAllocator;
> import java.lang.foreign.ValueLayout;
13c16
< import java.nio.file.Files;
---
> import java.io.File;
15,16c18,20
< import java.util.LinkedList;
< import java.util.List;
---
> import java.nio.charset.StandardCharsets;
> import java.util.Arrays;
> import java.util.Optional;
18a23,26
> import java.lang.foreign.AddressLayout;
> import java.lang.foreign.MemoryLayout;
> 
> import static java.lang.foreign.Linker.*;
28,35c36
<     private static boolean isLoaded = false;
<     private static boolean ayatana = false;
<     private static boolean appindicator = false;
<     private static final String LD_CONFIG = "/etc/ld.so.conf.d/";
<     private static final String APPINDICATOR_VERSION = "libappindicator3.so.1";
<     private static final String AYATANA_APPINDICATOR_VERSION = "libayatana-appindicator3.so.1";
<     private static  List<String> allPath = new LinkedList<>();
<     private static final Logger LOG = LoggerFactory.getLogger(RuntimeHelper.class);
---
>     static final AddressLayout POINTER = ValueLayout.ADDRESS.withTargetLayout(MemoryLayout.sequenceLayout(JAVA_BYTE));
38c39
<             (size, align) -> MemorySegment.allocateNative(size, align, SegmentScope.auto());
---
>             (size, align) -> Arena.ofAuto().allocate(size, align);
41,71c42
<         try (Stream<Path> paths = Files.list(Path.of(LD_CONFIG))) {
<             paths.forEach((file) -> {
<                 try (Stream<String> lines = Files.lines(file)) {
<                     List<String> collection = lines.filter(line -> line.startsWith("/")).toList();
<                     allPath.addAll(collection);
<                 } catch (IOException e) {
<                     LOG.error("File '{}' could not be loaded", file);
<                 }
<             });
<         } catch (IOException e) {
<             LOG.error("Directory '{}' does not exist", LD_CONFIG);
<         }
< 
<         allPath.add("/usr/lib"); // for systems, that don't implement multiarch
<         for (String path : allPath) {
<             try {
<                 System.load(path + File.separator + AYATANA_APPINDICATOR_VERSION);
<                 isLoaded = true;
<                 ayatana = true;
<                 break;
<             } catch (UnsatisfiedLinkError e) {
<                 try {
<                     System.load(path + File.separator + APPINDICATOR_VERSION);
<                     isLoaded = true;
<                     appindicator = true;
<                     break;
<                 } catch (UnsatisfiedLinkError ignored) { }
<             }
<         }
<         LOG.info(ayatana ? "Native code library " + AYATANA_APPINDICATOR_VERSION + " successfully loaded" : "Native code library " + AYATANA_APPINDICATOR_VERSION + " failed to load");
<         LOG.info(appindicator ? "Native code library " + APPINDICATOR_VERSION + " successfully loaded" : "Native code library " + APPINDICATOR_VERSION + " failed to load");
---
>         
87c58,60
<         return SYMBOL_LOOKUP.find(name).map(symbol -> MemorySegment.ofAddress(symbol.address(), layout.byteSize(), symbol.scope())).orElse(null);
---
>         return SYMBOL_LOOKUP.find(name)
>                 .map(s -> s.reinterpret(layout.byteSize()))
>                 .orElse(null);
114c87
<     static <Z> MemorySegment upcallStub(MethodHandle fiHandle, Z z, FunctionDescriptor fdesc, SegmentScope scope) {
---
>     static <Z> MemorySegment upcallStub(MethodHandle fiHandle, Z z, FunctionDescriptor fdesc, Arena scope) {
123,124c96,97
<     static MemorySegment asArray(MemorySegment addr, MemoryLayout layout, int numElements, SegmentScope scope) {
<          return MemorySegment.ofAddress(addr.address(), numElements * layout.byteSize(), scope);
---
>     static MemorySegment asArray(MemorySegment addr, MemoryLayout layout, int numElements, Arena arena) {
>          return addr.reinterpret(numElements * layout.byteSize(), arena, null);
271,275d243
<     }
< 
<     // Getter
<     public static boolean isLoaded() {
<         return isLoaded;
```

**Comment by purejava** [comment]
@overheadhunter changes are committed to the branches `minimal-jdk-21`, `develop-jdk-21` and `libayatana-appindicator-full-jdk-21` respectively.

Unfortunately, I cannot test the changes, as IntellliJ refuses to run 21.ea.35 with:
```
Fehler: Hauptklasse com.purejava.fxdemo.HelloApplication kann nicht in Modul com.purejava.fxdemo geladen werden
  java.lang.UnsupportedClassVersionError: Preview features are not enabled for com/purejava/fxdemo/HelloApplication (class file version 65.65535). Try running with '--enable-preview'

Process finished with exit code 1
```

I'll have to wait for the release of Java 21 in a couple of days.

**Comment by overheadhunter** [comment]
> `Try running with '--enable-preview'`

Actually, you can add this flag to the IntelliJ Run Config as well: :wink:

<img width="1152" alt="Bildschirmfoto 2023-09-17 um 10 05 00" src="https://github.com/purejava/appindicator-gtk3-java/assets/1204330/44207ad7-ce37-495f-85b7-86f6de8555ee">

**Comment by purejava** [comment]
Thanks. Yeah, I already have this set, and in the compiler options as well.

**Comment by purejava** [comment]
Tested and released with appindicator-gtk3-java release `1.3.6`.

Depends on https://github.com/cryptomator/integrations-linux/pull/29.
